kernel/src/sched.rs

```diff
@@ -332,69 +436,117 @@ impl Kernel {
         }
     }
 
-    /// Main loop.
-    pub fn kernel_loop<P: Platform, C: Chip>(
-        &'static self,
+    /// Main loop of the OS
+    pub fn kernel_loop<P: Platform, C: Chip, SC: Scheduler<C>>(
+        &self,
         platform: &P,
         chip: &C,
         ipc: Option<&ipc::IPC>,
+        scheduler: &SC,
         _capability: &dyn capabilities::MainLoopCapability,
-    ) {
+    ) -> ! {
         chip.watchdog().setup();
         loop {
             chip.watchdog().tickle();
             unsafe {
-                chip.service_pending_interrupts();
-                DynamicDeferredCall::call_global_instance_while(|| !chip.has_pending_interrupts());
-
-                for p in self.processes.iter() {
-                    chip.watchdog().tickle();
-                    p.map(|process| {
-                        self.do_process(platform, chip, process, ipc);
-                    });
-                    if chip.has_pending_interrupts()
-                        || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
-                    {
-                        break;
+                match scheduler.kernel_work_ready(chip) {
+                    true => {
+                        // Execute kernel work
+                        scheduler.execute_kernel_work(chip);
                     }
-                }
-
-                chip.atomic(|| {
-                    if !chip.has_pending_interrupts()
-                        && !DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
-                        && self.processes_blocked()
-                    {
-                        chip.watchdog().suspend();
-                        chip.sleep();
-                        chip.watchdog().resume();
+                    false => {
+                        // No kernel work ready, so ask scheduler for a process
+                        match scheduler.next(self) {
+                            SchedulingDecision::RunProcess((appid, timeslice_us)) => {
+                                self.process_map_or((), appid, |process| {
+                                    let (reason, time_executed) = self.do_process(
+                                        platform,
+                                        chip,
+                                        scheduler,
+                                        process,
+                                        ipc,
+                                        timeslice_us,
+                                    );
+                                    scheduler.result(reason, time_executed);
+                                });
```

Comment:
I think this is just a rust thing we have to deal with. A buggy scheduler that returns non existent `AppId`s doesn't seem like a case worth checking for.

Reply:
Yeah generally agreed this is another case for #1834 . We could add a `process_map_or_else()` and then panic in the missing case, I guess?

Reply:
That would protect against two cases:
1. The process moves in the array, or is removed, or otherwise stops existing or
2. The scheduler just made up an AppId.

The first is why process_map_or exists, but how could something happen to a process between the scheduler choosing from the list and the next line of code? And the second option could happen, I guess, but why would we add a panic!() output? It should be pretty obvious that the scheduler is calling `AppId::new()` and the scheduler doesn't work.

Reply:
yeah, lets just leave it. The only reason to change it is that the `map_or((), ...)` is a bit of a code smell, but there is an actual justification here.